docs: add changelog for 2.0.0 (beta.1) (#2528)
jupyterhub/zero-to-jupyterhub-k8s: docs/source/changelog.md

requests](https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pulls?q=is%3Apr+is%3Aclosed+label%3Abreaking),
this list should be updated.

## 2.0

### 2.0.0

#### Highlights

> - **TODO**
> 
>   TODO

Reviewer: ```suggestion
Z2JH 2.0.0 is the first major release since 1.0.0 was released in June 2021, and contains major upgrades to all JupyterHub components.

JupyterHub 2.x includes new RBAC support allowing fine-grained access control to hub services and servers

JupyterLab, the next-generation Notebook interface, is now the default interface seen by users. This brings a full development environment with a large number of extensions developed by the Jupyter community.

This release also includes several smaller changes that help Z2JH interface better with the rest of the Jupyter community such as not overriding a Docker image's command, and using standard Helm chart parameter names to match with other Helm charts.
Although these are breaking changes they will greatly improve the maintainability of the JupyterHub chart in future, and should also make it easier for new users to get started.
```

Thought  might as well make a start now.

`proxy.traefik` or `singleuser` to restrict the permissions to establish
outbound network connections, this upgrade is likely to escalate those
permissions unless you revise your configuration. The new configuration
`...networkPolicy.egressAllowRules` are by default granting most of the egress
permissions previously granted by default via the `...networkPolicy.egress`
configuration, and `...networkPolicy.egress` are now by default not providing
any permissions.

If you for example had overridden the previously very permissive default value
of `singleuser.networkPolicy.egress` to be a bit less permissive, you should if
you are unsure disable all `singleuser.networkPolicy.egressAllowRules` like this
to not risk escalating the permissions.

```yaml
singleuser:
  networkPolicy:
    egressAllowRules:
      cloudMetadataServer: false
      dnsPortsPrivateIPs: false
      nonPrivateIPs: false
      privateIPs: false
```

For more details, see the documentation on [Kubernetes Network
Policies](https://zero-to-jupyterhub.readthedocs.io/en/latest/administrator/security.html#kubernetes-network-policies)
and the configuration reference entries under

> [`...networkPolicy.egress`](https://zero-to-jupyterhub.readthedocs.io/en/latest/resources/reference.html#hub-networkpolicy-egress)

Author: Yes, as of #2769! I'll fix this now!

> [`...networkPolicy.egress`](https://zero-to-jupyterhub.readthedocs.io/en/latest/resources/reference.html#hub-networkpolicy-egress)

Author: Done!

and
[`...networkPolicy.egressAllowRules`](https://zero-to-jupyterhub.readthedocs.io/en/latest/resources/reference.html#hub-networkpolicy-egressallowrules).

#### Upgrade instructions

1. Adjust your configuration based on the security notice above if it applies to
   your deployment.
2. Prepare a user image with a modern version of JupyterHub installed. Install
   `jupyterhub>=2.3.1` via pip or `jupyterhub-base>=2.3.1` via conda/mamba in
   the image. Configure to use this new image, which is typically done via
   `singleuser.image` of as part of `singleuser.profileList`.
3. Read and adjust your configuration to the breaking changes below.

#### Breaking changes

- The following PRs have breaking labels since 1.1.3 release (1.1.4 and 1.2.0 are backport prs)

>   - use image for singleuser.cmd by default - https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/2449

Reviewer: ````suggestion

##### Default to using the container image's command instead of `jupyterhub-singleuser` https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/2449
Z2JH now launches the container's default command (equivalent to setting `CMD` in a `Dockerfile`) instead of overriding it. To restore the old bvehaviour set
```yaml
singleuser:
  cmd: jupyterhub-singleuser
```
````
I think it's cleaner to break these into separate subheadings instead of putting them all in a list, since they may contain a lot of context, but feel free to keep them as a list if you prefer.

  - match config load priority for jupyterhub_config.d files and hub.extraConfig https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/2457

>   - netpol stuff https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/2508
>     - users with egress to be warned about egressAllowRules opening up things by
>       default if they were limiting things before

Reviewer: ```suggestion

##### NetworkPolicy egress rules have been extended with a new property https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/2508
See `Security: breaking change to *.networkPolicy.egress`
```
Need to figure out how to cross-link this to the above section

  - userscheduler stuff https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/2590
  - require k8s 1.20+ https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/2635

>   - hub.fsGid replaced https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/2720

Reviewer: ````suggestion

##### `hub.fsGid` is replaced by `hub.podSecurityContext` https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/2720
In previous versions of Z2JH `hub.fsGid` set a supplemental group ID, which is required on some K8s systems to ensure JupyterHub has permissions to read/write files on a volume.
This has been replaced by the more general [`hub.podSecurityContext`](https://zero-to-jupyterhub.readthedocs.io/en/latest/resources/reference.html#hub-podsecuritycontext).
To upgrade set:
```yaml
hub:
  podSecurityContext:
    fsGroup: GROUP-ID
```
````

permissions previously granted by default via the `...networkPolicy.egress`
configuration, and `...networkPolicy.egress` are now by default not providing
any permissions.

If you for example had overridden the previously very permissive default value
of `singleuser.networkPolicy.egress` to be less permissive, you should consider
disabling all `singleuser.networkPolicy.egressAllowRules` like this
to not risk escalating the permissions.

```yaml
singleuser:
  networkPolicy:
    egressAllowRules:
      cloudMetadataServer: false
      dnsPortsPrivateIPs: false
      nonPrivateIPs: false
      privateIPs: false
```

For more details, see the documentation on [Kubernetes Network Policies](netpol)
and the configuration reference entries under
[`...networkPolicy.egress`](schema_hub.networkPolicy.egress) and
[`...networkPolicy.egressAllowRules`](schema_hub.networkPolicy.egressAllowRules).

#### Upgrade instructions

1. Adjust your configuration based on the security notice above if it applies to
   your deployment.
2. Prepare a user image with a modern version of JupyterHub installed. Install
   `jupyterhub>=2.3.1` via pip or `jupyterhub-base>=2.3.1` via conda/mamba in

>    the image. Configure to use this new image, which is typically done via

Reviewer: ```suggestion
   the image.
   Update the configuration to use this new image, which is typically done via
```

   `singleuser.image` or as part of `singleuser.profileList`.
3. Read and adjust your configuration to the breaking changes below.

#### Breaking changes

These breaking changes have been made relative to the 1.\* series of Z2JH releases.

> 

Reviewer: ```suggestion

##### JupyterHub 2 and related hub components
Z2JH 2.0.0 upgrades JupyterHub to the 2.\* series, and also upgrades all hub components.
If you are using any custom JupyterHub services, addons, API integrations, or extra configuration, you should review the breaking changes in the
[JupyterHub 2.0.0 changelog](https://github.com/jupyterhub/jupyterhub/blob/2.3.1/docs/source/changelog.md#200).

JupyterHub 2 uses an updated database schema.
Z2JH 2.0.0 automatically handles the upgrade, but it will not be possible to downgrade to older releases after this.

JupyterHub 2 adds RBAC for managing permissions in JupyterHub.
The old permissions model of admin/non-admin still works but you should use [RBAC to assign the required privileges to users or services in future](https://jupyterhub.readthedocs.io/en/stable/rbac/index.html)

See 
`TODO: link to Notable dependencies updated`
for more information on other upgraded hub components.

##### JupyterLab and Jupyter Server
The default singleuser server is [JupyterLab](https://jupyterlab.readthedocs.io/), running on [Jupyter server](https://jupyter-server.readthedocs.io/en/latest/).
To switch back to Jupyter Notebook either configure/rebuild your singleuser image to default to notebook, or see [the documentation on user interfaces](user-interfaces)

```
The anchors in the JupyterHub changelog on RTD aren't constant across versions, e.g. `https://jupyterhub.readthedocs.io/en/stable/changelog.html#id41`, so I've linked to the GitHub changelog instead.

```yaml
singleuser:
  cmd: jupyterhub-singleuser
```

- match config load priority for jupyterhub_config.d files and hub.extraConfig https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/2457

##### NetworkPolicy egress rules have been extended with a new property https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/2508

See `Security: breaking change to *.networkPolicy.egress`

- userscheduler stuff https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/2590
- require k8s 1.20+ https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/2635

##### `hub.fsGid` is replaced by `hub.podSecurityContext` https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/2720

In previous versions of Z2JH `hub.fsGid` set a supplemental group ID, which is required on some K8s systems to ensure JupyterHub has permissions to read/write files on a volume.
This has been replaced by the more general [`hub.podSecurityContext`](schema_hub.podSecurityContext).
To upgrade set:

```yaml
hub:
  podSecurityContext:
    fsGroup: GROUP-ID
```

- hub image rework to debian https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/2733

> - serviceAccount.create|name and rbac.create https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/2736, https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/2739

Reviewer: ```suggestion

##### Disabling RBAC requires setting multiple properties,`rbac.enable` is removed [#2736](https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/2736) [#2739](https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/2739)
If you previously disabled RBAC using `rbac.enable: False` you should set

```yaml
rbac:
  create: False
hub:
  serviceAccount:
    create: false
proxy:
  traefik:
    serviceAccount:
      create: false
scheduling:
  userScheduler:
    serviceAccount:
      create: false
prePuller:
  hook:
    serviceAccount:
      create: false
```

No changes are required if you previously used the default value of `rbac.enable: True`.
```

requests](https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pulls?q=is%3Apr+is%3Aclosed+label%3Abreaking),
this list should be updated.

## 2.0

### 2.0.0

#### Highlights

- **TODO**

>   TODO

Reviewer 2: What else should we highlight? It seems that most of the information below before the PR list _is_ the highlights, so maybe we should omit this? Or increase level so everything is h5 under the highlights h4?

Reviewer: I think a short paragraph rather than just bullet points would be nice to have here. I'll work on it unless someone else gets there first.

[`...networkPolicy.egress`](schema_hub.networkPolicy.egress) and
[`...networkPolicy.egressAllowRules`](schema_hub.networkPolicy.egressAllowRules).

#### Upgrade instructions

1. Adjust your configuration based on the security notice above if it applies to
   your deployment.
2. Prepare a user image with a modern version of JupyterHub installed. Install
   `jupyterhub>=2.3.1` via pip or `jupyterhub-base>=2.3.1` via conda/mamba in
   the image.
   Update the configuration to use this new image, which is typically done via
   `singleuser.image` or as part of `singleuser.profileList`.
3. Read and adjust your configuration to the breaking changes below.

#### Breaking changes

These breaking changes have been made relative to the 1.\* series of Z2JH releases.

##### JupyterHub 2 and related hub components

Z2JH 2.0.0 upgrades JupyterHub to the 2.\* series, and also upgrades all hub components.
If you are using any custom JupyterHub services, addons, API integrations, or extra configuration, you should review the breaking changes in the
[JupyterHub 2.0.0 changelog](https://github.com/jupyterhub/jupyterhub/blob/2.3.1/docs/source/changelog.md#200).

> JupyterHub 2 uses an updated database schema.
> Z2JH 2.0.0 automatically handles the upgrade, but it will not be possible to downgrade to older releases after this.

Reviewer 2: ```suggestion
JupyterHub 2 updates the database schema, which means a migration takes place when you upgrade JupyterHub.
Z2JH automatically handles the upgrade if you are using sqlite (`hub.db.type = 'sqlite-pvc', the default), but it may not be possible to downgrade to older releases after this.
When using sqlite, JupyterHub automatically creates a backup in the `hub-db` volume,
which can be restored manually if you need to downgrade.
```

Reviewer: @minrk I've copied your suggestions to the new upgade guide: https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/2803

JupyterHub 2 adds RBAC for managing permissions in JupyterHub.

> The old permissions model of admin/non-admin still works but you should use [RBAC to assign the required privileges to users or services in future](https://jupyterhub.readthedocs.io/en/stable/rbac/index.html)

Reviewer 2: ```suggestion
The old permissions model of admin/non-admin still works, but we recommend using [RBAC to assign only the required privileges to users or services in future](https://jupyterhub.readthedocs.io/en/stable/rbac/index.html)

Default permissions are mostly unchanged, but a few have:

- Servers' own API tokens have limited permissions by default, which can be expanded by defining the `server` role. The previous behavior was the maximum permission of `inherit`.
- `admin_access` as a concept is removed, so disabling it has no effect. In 2.0, admins by definition can do everything, including access servers. To limit user permissions, assign them to roles which have only the needed permissions.
```

> See
> `TODO: link to Notable dependencies updated`
> for more information on other upgraded hub components.

Reviewer 2: ```suggestion
See
[Notable dependencies updated](notable-dependencies-200)
for more information on other upgraded hub components.
```

##### JupyterLab and Jupyter Server

The default singleuser server is [JupyterLab](https://jupyterlab.readthedocs.io/), running on [Jupyter server](https://jupyter-server.readthedocs.io/en/latest/).
To switch back to Jupyter Notebook either configure/rebuild your singleuser image to default to notebook, or see [the documentation on user interfaces](user-interfaces)

##### Default to using the container image's command instead of `jupyterhub-singleuser` [#2449](https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/2449)

> Z2JH now launches the container's default command (equivalent to setting `CMD` in a `Dockerfile`) instead of overriding it. To restore the old bvehaviour set

Reviewer 2: ```suggestion
Z2JH now launches the container's default command (as set e.g. by `CMD` in a `Dockerfile`) instead of overriding it. To restore the old behaviour, set
```

Note the singleuser image is not affected.

##### Disabling RBAC requires setting multiple properties,`rbac.enable` is removed [#2736](https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/2736) [#2739](https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/2739)

If you previously disabled RBAC using `rbac.enable: False` you should set

````yaml
rbac:
  create: False
hub:
  serviceAccount:
    create: false
proxy:
  traefik:
    serviceAccount:
      create: false
scheduling:
  userScheduler:
    serviceAccount:
      create: false
prePuller:
  hook:
    serviceAccount:
      create: false

- The following breaking changes are relevant by bumping various packages:
  - kubespawner
    - kubernetes_async
    - sudo config
    - not mounting serviceaccount by default?
  - oauthenticator
  - jupyterhub
    - RBAC permissions, so granting admin rights to services isn't recommended etc.
    - database schema upgrade needed

#### TODO: link out to relevant informational posts

TODO: Write a discourse forum post about the following and highlight that via

> the changelog.

Reviewer 2: What's the order for this? Write the post now, then link here, or make a first beta, then write the post, then add link here?

Reviewer: Good point, I think getting the beta out first with a link to the changelog and upgrade guide makes sense and will help speed things up. Save the discourse post (and maybe something on the official Jupyter blog) for the final release?

Author: ```
#### TODO: link out to relevant informational posts

TODO: Write a discourse forum post about the following and highlight that via
the changelog.

- 100+ user-placeholder pods -> node-placeholder pods
- autohttps acme acquisition issues on GKE mitigation
  (notable-dependencies-200)=
```

I think it is quite crucial to be aware of that user placeholder pods shouldn't be over 100 or similar, and that there are currently some issues on GKE with slow setup of networking - which causes failures in the autohttps system. But, lets get this out there now instead and hope that we manage to communicate that sometime else instead.

In [e8cc8a1](https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/2528/commits/e8cc8a1a141f1e442d690ec7dfb420378f0556e4) I removed this TODO entry.

- 100+ user-placeholder pods -> node-placeholder pods
- autohttps acme acquisition issues on GKE mitigation

> 

Reviewer 2: Adding anchor target

```suggestion
(notable-dependencies-200)=

```
